<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Allele-specific copy number genotyping for single cell
sequencing data</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Allele-specific copy number genotyping for single cell
sequencing data
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;Alleloscope&rsquo; version 1.0.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="AssignClones_ref.html">AssignClones_ref</a></td>
<td>Using marker regions to assign each cell into c reference subclones</td></tr>
<tr><td style="width: 25%;"><a href="Createobj.html">Createobj</a></td>
<td>Generate Alleloscope object for analysis</td></tr>
<tr><td style="width: 25%;"><a href="EM.html">EM</a></td>
<td>Iterative phasing and theta_hat estimation</td></tr>
<tr><td style="width: 25%;"><a href="Est_regions.html">Est_regions</a></td>
<td>Perform EM iterations on the filtered cells with barcodes, and plot the results for each region.</td></tr>
<tr><td style="width: 25%;"><a href="Genotype.html">Genotype</a></td>
<td>Genotype each cell for each region and plot the genotypes.</td></tr>
<tr><td style="width: 25%;"><a href="Genotype_value.html">Genotype_value</a></td>
<td>Normalize coverage using identified/ specified normal cells and one normal region and generate a table with (rho_hat, theta_hat) of each cell for all regions.</td></tr>
<tr><td style="width: 25%;"><a href="Lineage_plot.html">Lineage_plot</a></td>
<td>Generate genotype plot (scatter plot) for each region and save in the plot directory.</td></tr>
<tr><td style="width: 25%;"><a href="Matrix_filter.html">Matrix_filter</a></td>
<td>Filter object based on cell number for each SNP, SNP number for each cell, SNP variant allele frequency, and exclude the centromere and telomere regions.</td></tr>
<tr><td style="width: 25%;"><a href="Rundf_dna.html">Rundf_dna</a></td>
<td>Run all steps for scDNA-seq data</td></tr>
<tr><td style="width: 25%;"><a href="Segmentation.html">Segmentation</a></td>
<td>HMM segmentation based on coverage matrix for paired tumor and normal sample.</td></tr>
<tr><td style="width: 25%;"><a href="Segments_filter.html">Segments_filter</a></td>
<td>Select the segments in the "seg_table" with more than nSNP</td></tr>
<tr><td style="width: 25%;"><a href="Select_normal.html">Select_normal</a></td>
<td>Identify candidate normal cells and normal regions for cell coverage normalization</td></tr>
</table>
</body></html>
